bpf/cpu/rbperf.bpf.c

```diff
@@ -0,0 +1,438 @@
+            state->stack.frames.len += 1;
+        }
+
+        cfp += control_frame_t_sizeof;
+    }
+
+    state->cfp = cfp;
+    state->base_stack = base_stack;
+
+    if (cfp <= base_stack &&
+        state->ruby_stack_program_count < BPF_PROGRAMS_COUNT) {
+        LOG("[debug] traversing next chunk of the stack in a tail call");
+        bpf_tail_call(ctx, &programs, RBPERF_STACK_READING_PROGRAM_IDX);
+    }
+
+    state->stack.stack_status = cfp > state->base_stack ? STACK_COMPLETE : STACK_INCOMPLETE;
+
+    if (state->stack.frames.len != state->stack.expected_size) {
+        LOG("[error] stack size %d, expected %d", state->stack.frames.len, state->stack.expected_size);
+    }
+
+    // Hash stack.
+    int ruby_stack_hash = MurmurHash2((u32 *)state->stack.frames.addresses, MAX_STACK * sizeof(u64) / sizeof(u32), 0);
+    LOG("[debug] ruby stack hash: %d", ruby_stack_hash);
+
+    unwind_state_t *unwind_state = bpf_map_lookup_elem(&heap, &zero);
+    if (unwind_state != NULL) {
+        unwind_state->stack_key.interpreter_stack_id = ruby_stack_hash;
+    }
+
+    // Insert stack.
+    int err = bpf_map_update_elem(&interpreter_stack_traces, &ruby_stack_hash, &state->stack.frames, BPF_ANY);
+    if (err != 0) {
+      LOG("[error] bpf_map_update_elem with ret: %d", err);
+    }
+
+    // We are done.
+    aggregate_stacks();
+    return 0;
+}
+
+SEC("perf_event")
+int unwind_ruby_stack(struct bpf_perf_event_data *ctx) {
+    u64 zero = 0;
+    unwind_state_t *unwind_state = bpf_map_lookup_elem(&heap, &zero);
+    if (unwind_state == NULL) {
+        bpf_printk("[rbperf] unwind_state is NULL, should not happen");
+        return 1;
+    }
+    // bpf_printk("[rbperf] unwind_state->len = %d", unwind_state->stack.len);
+
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    // There's no point in checking for the swapper process.
+    if (pid == 0) {
+        return 0;
+    }
+
+    ProcessData *process_data = bpf_map_lookup_elem(&pid_to_rb_thread, &pid);
+
+    if (process_data != NULL && process_data->rb_frame_addr != 0) {
+        struct task_struct *task = (void *)bpf_get_current_task();
+        if (task == NULL) {
+            LOG("[error] task_struct was NULL");
+            return 0;
+        }
+
+        // PIDs in Linux are reused. To ensure that the process we are
+        // profiling is the one we expect, we check the pid + start_time
+        // of the process.
+        //
+        // When we start profiling, the start_time will be zero, so we set
+        // it to the actual start time. Otherwise, we check that the start_time
+        // of the process matches what we expect. If it's not the case, bail out
+        // early, to avoid profiling the wrong process.
+        if (enable_pid_race_detector) {
+            u64 process_start_time;
+            bpf_core_read(&process_start_time, 8, &task->start_time);
```

Comment:
Do we provide this in the parca agent user-space code?

Reply:
No because this feature is currently disabled https://github.com/parca-dev/parca-agent/blob/e5a1dc8e844d5cd19159471e745a8f63d1ef7e17/bpf/cpu/rbperf.bpf.c#L72. If we decide to enable this in the Agent (we should do this in the entry-point program, though) we should do it in another PR I think